aten/src/ATen/core/ITensorList.h

```diff
@@ -0,0 +1,367 @@
+#include <c10/util/Exception.h>
+
+#include <initializer_list>
+#include <iterator>
+#include <type_traits>
+
+namespace at {
+class Tensor;
+}
+
+namespace c10 {
+class ITensorList;
+class ITensorListIterator;
+
+/*
+ * Adding a new tag
+ * ================
+
+ * -> Short Version
+ *
+ * If you are lucky, the only necessary things for introducing
+ * a new tag are:
+ *
+ *     - Add a line to `TORCH_ITENSORLIST_FORALL_TAGS` such as:
+ *                 `_(TagName, ##__VA_ARGS__)`
+ *
+ *     - Specialize `ITensorListTagImpl` to your tag, implementing
+ *       the required methods
+ *
+ *     - Implement constructors for the type corresponding to your
+ *       tag on both `ITensorList(Iterator)` classes
+ *
+ * -> Long Version
+ *
+ * That may not work if your container is "more different" than
+ * the ones implemented. Specifically, maybe, some of the functions
+ * implemented with `TORCH_ITENSORLIST_UNWRAP` may not work for it.
+ * In those cases, you should do something like `get` and
+ * `iterator_get` functions.
+ *
+ * Say the implementation for `begin` is different. Then, you should:
+ *
+ *     - Do everything done in the "Short Version" for introducing
+ *       the new tag
+ *
+ *     - Implement a new function `begin` in all implementation
+ *       classes (for all tag types)
+ *
+ *     - Call it directly inside a `TORCH_ITENSORLIST_UNWRAP`
+ */
+
+// Applies arbitrary macros to each `ITensorListTag`.
+#define TORCH_ITENSORLIST_FORALL_TAGS(_, ...) \
+  _(Unboxed, ##__VA_ARGS__)                   \
+  _(Boxed, ##__VA_ARGS__)
+
+// Builds the name of the implementation class for `TAG`.
+#define TORCH_ITENSORLIST_IMPL(TAG) \
+  c10::detail::ITensorListTagImpl<c10::ITensorListTag::TAG>
+
+// Defines a "switch-case" for `TAG`. Inside, it executes `BODY`,
+// while bringing to scope:
+//     - `ImplT`: the implementation class for `TAG`
+//     - `this_`: the result of unwrapping `this`
+#define TORCH_ITENSORLIST_UNWRAP_CASE(TAG, BODY) \
+  case c10::ITensorListTag::TAG: {               \
+    using ImplT = TORCH_ITENSORLIST_IMPL(TAG);   \
+    auto& this_ = ImplT::unwrap(*this);          \
+    BODY                                         \
+  } break;
+
+// Dispatches the unwrap call, depending on `TAG`, followed by
+// the execution of `BODY`. It aborts if `TAG` is not a `ITensorListTag`.
+#define TORCH_ITENSORLIST_UNWRAP(TAG, BODY)                            \
+  switch (TAG) {                                                       \
+    TORCH_ITENSORLIST_FORALL_TAGS(TORCH_ITENSORLIST_UNWRAP_CASE, BODY) \
+    default:                                                           \
+      TORCH_INTERNAL_ASSERT(false, "invalid ITensorList tag.");        \
+  }
+
+enum class ITensorListTag {
+#define DEFINE_TAG(tag, ...) tag,
+  TORCH_ITENSORLIST_FORALL_TAGS(DEFINE_TAG)
+#undef DEFINE_TAG
+      None
+};
+
+namespace detail {
+using ITensorListConstRef =
+    typename detail::ivalue_to_const_ref_overload_return<at::Tensor>::type;
```

Comment:
Is there a downside to killing this and just using `const Tensor&` where you previously had `ITensorListConstRef`? That might be a little easier to read.

Reply:
There's one: I will have to introduce it again when supporting `Tensor?[]`.